### DISCUSSION TITLE
A simpler bunshi?

### DISCUSSION THREAD

**Discussion by yf-yang** [pmndrs/jotai#2217] [Ideas]
This one is a little bit off topic, but since bunshi comes from jotai, I'll discuss the idea here first.

When using jotai-scope, I found I actually can divide my needs into two categories, which represents two use cases of React.Context. In a specific scope, 
- if the context value can change, that's a state management problem —— jotai-scope
- if the context value never changes, then I use React.Context just to avoid passing a prop component by component, that's the dependency injection problem.

For dependency injection, I have three choices:
- Original React.Context. I found that I don't like it because I need to write things like `<MyContext.Provider value={}>` instead of `<UnifiedProvider scope={} value={}>`.
- jotai-scope. Actually I do not mean to use it for that case, but just now I realize I am using jotai-scope for dependency injection. However, I pay for extra costs for that misuse. Listeners are added but they are never called because the atoms acting as dependency injection tool is not stateful at all.
- So I think about bunshi. However, I don't like the molecule-molecule dependency design because I really can't get the point and cannot come up with any use case.

Therefore, I am thinking about a set of api that performs dependency injection:
``` javascript
// the default dependency being injected
const dep;
const scopeAtom = atom( dep );
<ScopeProvider atom={scopeAtom} value={anotherDep}></ScopeProvider>
```
Something like that, then rename atom to scope. It's like both simplified jotai-scope (without those listeners to sync state change because there is no state) and simplified bunshi (without the idea of molecule and molecule-molecule dependency, now useMolecule becomes useScope, something equivalent to useAtomValue).

**Comment by dai-shi** [comment]
I'm not sure if I get the idea fully.

> Original React.Context.

If that works, wouldn't it be the perfect solution?
You could build `UnifiedProvider` without Jotai, no?
(So, that's probably why you mean by "a simpler bunshi", independent from Jotai.)

**Reply by yf-yang** [reply]
Yes, this is not related to jotai or future jotai feature.
I know React.Context is the less overhead one, but from jotai's perspective, I would like to have similar APIs.
For React, different dependencies have different providers. For something similar to jotai style APIs, it should be unified provider with different context as one of the input props.
Anyway, it is not so hard to implement one.

This topic is actually related to one of a broader topic: When replacing React useState/useReducer/useContext hooks with jotai, which pattern should be mapped to which pattern. For those simple dependency injection cases, jotai is too heavy (useless listeners), so this idea, together with Jotai, forms a set of similar API, which replaces those three React APIs. I may illustrate my classification of React common patterns in a future doc.

**Reply by dai-shi** [reply]
fwiw, one of my works is [use-context-selector](https://github.com/dai-shi/use-context-selector), which just does the job like UnifiedProvider (but with subscriptions), and then developed [use-atom](https://github.com/dai-shi/use-atom), which turned into Jotai. (And, re-implemented `use-atom` afterwards.)

**Reply by yf-yang** [reply]
Tried to write an extremely simple version:
``` typescript
import type { PropsWithChildren } from 'react';
import { createContext, useContext, useEffect, useRef } from 'react';

type UnknownDep = Dep<unknown>;

const DepsContext = createContext(new Map<UnknownDep, unknown>());

export interface Dep<T> {
  defaultValue: T;
  _depBrand: undefined;
}

export function dep<T>(defaultValue?: T): Dep<T> {
  return { defaultValue } as Dep<T>;
}

export function DepProvider<T>({
  dep,
  value,
  children,
}: PropsWithChildren<{
  dep: Dep<T>;
  value: T;
}>): JSX.Element {
  if (process.env.NODE_ENV !== 'production') {
    // eslint-disable-next-line react-hooks/rules-of-hooks -- Only work in dev mode
    const isFirstRender = useRef(true);

    if (!isFirstRender.current) {
      logger.warn(`DepProvider re-renders unexpectedly.
Prop dep should not change.
Prop value should not change, or it should be an atom in ScopeProvider.
If both of them do not change but this component re-renders, consider wrap DepProvider with React.memo`);
    }

    // HACK: https://stackoverflow.com/a/76242764/5930115
    // eslint-disable-next-line react-hooks/rules-of-hooks -- Only work in dev mode
    useEffect(() => {
      if (isFirstRender.current) {
        isFirstRender.current = false;
      }
    }, []);
  }
  const parentDeps = useContext(DepsContext);
  const deps = new Map([...parentDeps, [dep, value]]);
  return <DepsContext.Provider value={deps}>{children}</DepsContext.Provider>;
}

export function useDep<T>(dep: Dep<T>): T {
  const deps = useContext(DepsContext);
  return deps.has(dep) ? (deps.get(dep) as T) : dep.defaultValue;
}
```
